api/src/main/java/org/openmrs/api/impl/LocalStorageService.java

```diff
@@ -0,0 +1,158 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.impl;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Stream;
+
+import org.openmrs.api.StorageService;
+import org.openmrs.util.OpenmrsUtil;
+import org.springframework.beans.factory.annotation.Value;
+
+/**
+ * Used to persist data in a local file system or volumes.
+ * <p>
+ * It is the default implementation of StorageService.
+ */
+public class LocalStorageService extends BaseOpenmrsService implements StorageService {
+	
+	private final String storageDir;
+	
+	private final DateFormat dateFormat = new SimpleDateFormat("yyyy/MM-dd/HH-mm/yyyy-MM-dd-HH-mm-ss-SSS-");
+	
+	public LocalStorageService() {
+		this.storageDir = Paths.get(OpenmrsUtil.getApplicationDataDirectory(), "storage").toString();
+	}
+	
+	public LocalStorageService(@Value("${STORAGE_DIR}") String storageDir) {
+		this.storageDir = storageDir;
+	}
+	
+	@Override
+	public InputStream getData(final String key) throws IOException {
+		return Files.newInputStream(getPath(key));
+	}
+	
+	public Path getPath(final String key) {
+		Path path = Paths.get(key);
+		if (path.isAbsolute()) {
+			return path;
+		} else {
+			return Paths.get(storageDir, key);
+		}
+	}
+	
+	@Override
+	public Stream<String> getKeys(final String moduleId, final String prefix) throws IOException {
+		List<String> dirs = new ArrayList<>();
+		final Path storagePath;
+		if (moduleId != null) {
+			dirs.add("modules");
+			dirs.add(moduleId);
+			storagePath = Paths.get(storageDir, "modules", moduleId);
+		} else {
+			storagePath = Paths.get(storageDir);
```

Comment:
If `moduleId` is `null`, should we filter the `modules` directory from the returned results? Or else are we creating the contract that every storage service should have a top-level key called `modules` that stores the module-specific data?

Reply:
I'll slightly change this in the next revision I push.